File: opentera_webrtc_ros/src/RosSignalingServerConfiguration.cpp
#include <rclcpp/rclcpp.hpp>
#include <opentera_webrtc_ros/RosSignalingServerConfiguration.h>
#include <opentera_webrtc_ros/RosStreamBridge.h>
#include <opentera_webrtc_ros/RosNodeParameters.h>

using namespace opentera;
using namespace std;
using namespace rclcpp;

std::string
    replaceProtocol(const std::string& url, const char* from0, const char* to0, const char* from1, const char* to1)
{
    size_t p = url.find("://");
    if (p == std::string::npos)
    {
        return url;
    }

    std::string protocol = url.substr(0, p);
    transform(protocol.begin(), protocol.end(), protocol.begin(), ptr_fun<int, int>(tolower));
    if (protocol == from0)
    {
        return to0 + url.substr(p);
    }
    else if (protocol == from1)
    {
        return to1 + url.substr(p);
    }
    else
    {
        return url;
    }
}

std::string httpToWs(const std::string& url)
{
    return replaceProtocol(url, "http", "ws", "https", "wss");
}

std::string wsToHttp(const std::string& url)
{
    return replaceProtocol(url, "ws", "http", "wss", "https");
}

/**
 * @brief Build a signaling server configuration from the ROS parameter server
 * Parameters are retrieved from the signaling namespace under the node private namespace
 *
 * @param defaultClientName Default name for the webrtc peer
 * @return The signaling server configuration
 */
SignalingServerConfiguration RosSignalingServerConfiguration::fromRosParam(RosNodeParameters& nodeParameters)
{
    string serverUrl, clientName, room, password;
    nodeParameters.loadSignalingParams(serverUrl, clientName, room, password);

    return SignalingServerConfiguration::create(httpToWs(serverUrl) + "/signaling", clientName, room, password);
}

/**
 * @brief Build a signaling server configuration from the given Url.
 * Parameters are retrieved from the queries and from the signaling namespace under the node private namespace.
 *
 * @param url An url to parse.
 *
 * @return The signaling server configuration.
 */
SignalingServerConfiguration
    RosSignalingServerConfiguration::fromUrl(RosNodeParameters& nodeParameters, const std::string& url)
{
    string address = httpToWs(RosSignalingServerConfiguration::getBaseUrl(nodeParameters, url)) + "/signaling";

    size_t pos1 = url.find("?");
    string queries = url.substr(pos1);
    string password = getQueryFrom("pwd", queries);
    string clientName, roomName;

    nodeParameters.loadSignalingParams(clientName, roomName);

    return SignalingServerConfiguration::create(address, clientName, roomName, password);
}


/**
 * @brief Get a query value among the queries arguments.
 * Queries format: "?query1=123&query2=abc"
 *
 * @param query The parameters that contain de value.
 * @param queries The full queries in which the query argument his.
 *
 * @return The value of the query
 */
std::string RosSignalingServerConfiguration::getQueryFrom(const std::string& query, const std::string& queries)
{
    size_t pos1, pos2;

    pos1 = queries.find(query + "=");
    string remaining = queries.substr(pos1);
    pos2 = remaining.find_first_of("&");

    return remaining.substr(0, pos2).substr(query.size() + 1);
}

/**
 * @brief Get the ice servers url from url
 *
 * @param url The full url to extract from.
 * @return std::string The ice server url.
 */
std::string RosSignalingServerConfiguration::getIceServerUrl(RosNodeParameters& nodeParameters, const std::string& url)
{
    RCLCPP_INFO_STREAM(nodeParameters.node().get_logger(), "getIceServerUrl from url:" << url);
    return wsToHttp(RosSignalingServerConfiguration::getBaseUrl(nodeParameters, url)) + "/iceservers";
}

/**
 * @brief Get base url from url
 *
 * @param url The full url to extract from.
 * @return std::string The base url (up to the last / of the path).
 */
std::string RosSignalingServerConfiguration::getBaseUrl(RosNodeParameters& nodeParameters, const std::string& url)
{
    const string prot_end("://");
    string::const_iterator prot_i = search(url.begin(), url.end(), prot_end.begin(), prot_end.end());

    std::string protocol;
    protocol.reserve(distance(url.begin(), prot_i));


    transform(url.begin(), prot_i, back_inserter(protocol), ptr_fun<int, int>(tolower));  // protocol is icase

    if (prot_i == url.end())
    {
        RCLCPP_ERROR_STREAM(nodeParameters.node().get_logger(), "No protocol defined in url: " << url);
        return url;
    }

    advance(prot_i, prot_end.length());

    string::const_iterator path_i = std::find(prot_i, url.end(), '/');

    auto pos = url.find_last_of('/');
    string::const_iterator last_i = url.begin();
    advance(last_i, pos);

    if (path_i < last_i)
    {
        path_i = last_i;
    }

    std::string host;
    host.reserve(distance(prot_i, path_i));
    transform(prot_i, path_i, back_inserter(host), ptr_fun<int, int>(tolower));  // host is icase

    return protocol + prot_end + host;
}

File: opentera_webrtc_rosbridge_server/src/opentera_webrtc_rosbridge_server/ros_signaling_server_configuration.py
from opentera_webrtc.native_client import SignalingServerConfiguration
from typing import Optional

import rospy
from urllib.parse import urlparse


def from_ros_param() -> SignalingServerConfiguration:
    signaling_params = rospy.get_param("~signaling", {})

    server_url = signaling_params.get(
        "server_url", "http://localhost:8080")
    client_name = signaling_params.get("client_name", "rosbridge_server")
    room = signaling_params.get("room", "rosbridge")
    password = signaling_params.get("password", "abc")

    return SignalingServerConfiguration.create(url=server_url, client_name=client_name, room=room, password=password)


def from_url(url: str) -> SignalingServerConfiguration:
    address = f"{get_base_url(url)}/socket.io"

    queries = _extract_queries(url)
    password = _get_query_from("pwd", queries) or ""

    signaling = rospy.get_param("~signaling", {})
    client_name = signaling.get("client_name", "rosbridge_server")
    room = signaling.get("room", "rosbridge")

    return SignalingServerConfiguration.create(url=address, client_name=client_name, room=room, password=password)


def _get_query_from(query: str, queries: str) -> Optional[str]:
    pos1 = queries.find(f"{query}=")

    if pos1 == -1:
        rospy.logerr(f"[{query}] not found in [{queries}]")
        return None

    remaining = queries[pos1:]
    pos2 = remaining.find("&")

    return remaining[len(query) + 1:pos2] if pos2 != -1 else remaining[len(query) + 1:]


def get_base_url(url: str) -> str:
    parsed_url = urlparse(url)

    if parsed_url.scheme is None:
        rospy.logerr(f"No protocol defined in url: {url}")
        return url

    return f"{parsed_url.scheme}://{parsed_url.netloc}"


def get_ice_server_url(url: str) -> str:
    rospy.loginfo(f"get_ice_server_url from url: {url}")
    return f"{get_base_url(url)}/iceservers"


def _extract_queries(url: str) -> str:
    parsed_url = urlparse(url)

    if parsed_url.query is None:
        return "?"
    return f"?{parsed_url.query}"
